cmd/sonobuoy/app/gen_plugin_def.go

```diff
@@ -0,0 +1,146 @@
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/heptio/sonobuoy/pkg/errlog"
+	"github.com/heptio/sonobuoy/pkg/plugin/manifest"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	v1 "k8s.io/api/core/v1"
+	kuberuntime "k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	defaultPluginName    = "plugin"
+	defaultPluginDriver  = "Job"
+	defaultMountPath     = "/tmp/results"
+	defaultMountName     = "results"
+	defaultMountReadOnly = false
+)
+
+// GenPluginDefConfig are the input options for running
+type GenPluginDefConfig struct {
+	def manifest.Manifest
+
+	// env holds the values that will be parsed into the manifest env vars.
+	env EnvVars
+
+	// driver holds the values that will be parsed into the plugin driver.
+	// Allows validation during flag parsing by having a custom type.
+	driver pluginDriver
+}
+
+// NewCmdGenPluginDef ...
+func NewCmdGenPluginDef() *cobra.Command {
+	genPluginOpts := GenPluginDefConfig{
+		def: defaultManifest(),
+		env: map[string]string{},
+	}
+
+	cmd := &cobra.Command{
+		Use:   "plugin",
+		Short: "Generates the manifest Sonobuoy uses to define a plugin",
+		Run:   genPluginDefWrapper(&genPluginOpts),
+		Args:  cobra.NoArgs,
+	}
+
+	cmd.Flags().StringVarP(
+		&genPluginOpts.def.SonobuoyConfig.PluginName, "name", "n", "",
+		"Plugin name",
+	)
+	cmd.MarkFlagRequired("name")
+
+	cmd.Flags().VarP(
+		&genPluginOpts.driver, "type", "t",
+		"Plugin Driver (job or daemonset)",
+	)
+
+	cmd.Flags().StringVarP(
+		&genPluginOpts.def.Spec.Image, "image", "i", "",
+		"Plugin image",
+	)
+	cmd.MarkFlagRequired("image")
+
+	cmd.Flags().StringArrayVarP(
+		&genPluginOpts.def.Spec.Command, "cmd", "c", []string{"./run.sh"},
+		"Command to run when starting the plugin's container",
+	)
+
+	cmd.Flags().VarP(
+		&genPluginOpts.env, "env", "e",
+		"Env var values to set on the image (e.g. --env FOO=bar)",
+	)
+
+	return cmd
+}
+
+// defaultManifest returns the basic manifest the user's options will be placed
+// on top of.
+func defaultManifest() manifest.Manifest {
+	m := manifest.Manifest{}
+	m.Spec.Name = defaultPluginName
+	m.SonobuoyConfig.Driver = defaultPluginDriver
+	m.Spec.VolumeMounts = []v1.VolumeMount{
+		v1.VolumeMount{
+			MountPath: defaultMountPath,
+			Name:      defaultMountName,
+			ReadOnly:  defaultMountReadOnly,
+		},
+	}
+	return m
+}
+
+// genPluginDefWrapper returns a closure around a given *GenPluginDefConfig that
+// will adhere to the method signature needed by cobra. It prints the result of
+// the action or logs an error and exits with non-zero status.
+func genPluginDefWrapper(cfg *GenPluginDefConfig) func(cmd *cobra.Command, args []string) {
+	return func(cmd *cobra.Command, args []string) {
+		s, err := genPluginDef(cfg)
+		if err != nil {
+			errlog.LogError(err)
+			os.Exit(1)
+		}
+		fmt.Println(s)
+	}
+}
+
+// genPluginDef returns the YAML for the plugin which Sonobuoy would expect as
+// a configMap in order to run/gen a typical run.
+func genPluginDef(cfg *GenPluginDefConfig) (string, error) {
```

Comment:
I kind of expected this to have a return type of `([]byte, error)`, but I can see why this works as `(string, error)`